runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/DfsArtifactRetrievalService.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.fnexecution.artifact;
+
+
+import com.google.protobuf.ByteString;
+import io.grpc.stub.StreamObserver;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.ReadableByteChannel;
+import org.apache.beam.model.jobmanagement.v1.ArtifactApi;
+import org.apache.beam.model.jobmanagement.v1.ArtifactRetrievalServiceGrpc;
+import org.apache.beam.sdk.io.FileSystems;
+import org.apache.beam.sdk.io.fs.ResolveOptions;
+import org.apache.beam.sdk.io.fs.ResourceId;
+
+/**
+ * An {@link ArtifactRetrievalService} that uses distributed file systems as its backing storage.
```

Comment:
So if I understand correctly this works like this, please correct me if I'm wrong:

- It uses staging_token as the directory where the job's staging stuff is stored - both the artifacts and the manifest
- The manifest is stored in a file $token/MANIFEST
- Artifacts are stored in $token/$artifactName
- The manifest is expected to have .metadata.name be full paths to the artifacts, i.e. $token/$originalName

I would propose to change this a bit, considering that manifest should reflect what artifacts do we have rather than where they are stored; i.e. I think the name of the artifact in the manifest must be identical to its name in PutArtifactRequest, so it can not contain the staging token.

I suggest something like the following:

- Still use staging_token as the directory where the job's staging stuff is stored - both the artifacts and the manifest
- Still store manifest in $token/MANIFEST
- Manifest contains the original artifact names from PutArtifactRequest
- Have an "artifact map" file e.g. $token/MAP, which contains a mapping from original artifact name to path on the storage, e.g.:
```
{
  "foo.jar": "hdfs://path/to/staging/foo.jar",
  "some/data.pb": "hdfs://path/to/staging/some/data.pb",
  ...
}
```

Then when retrieving the artifact by name, you look it up in the map and download the file from where it points to.
This flexible mapping will also allow to later store files in something like path/to/$md5, allowing for deduplication of artifacts (avoiding uploading artifacts we already have).

Reply:
As discussed in the other PR, I agree that this is a far better approach.  I'll change this PR to integrate your suggestions.

Reply:
Suggestion: lets keep the $token as path to MANIFEST (including MANIFEST) in the end. This removes any implicit assumption on directory structure.

Also structure of MANIFEST is defined by proto https://github.com/apache/beam/blob/be6185fd5c2b471f371d7591108360974954ec12/model/job-management/src/main/proto/beam_artifact_api.proto#L74

Reply:
@angoenka I think I disagree.  If we do that, then it complicates the implementation of artifact retrieval because the storage location is no longer `$token/$artifactName`.  I'm not sure the gains from disambiguating manifest location justify the added complexity.

Reply:
Artifact location can never be assumed to be $token/$artifactName.

The artifacts are saved with a different name than the provided artifact name and the mapping of artifact name vs file name is stored in the manifest file.
The ProxyManifest is the proto which describe the format of manifest file. Its an optional format but it fits well in our use case.

By providing the base dir name as the token, the retrieval service assumes the folder structure (what i mean is it assumes that the manifest will be  $stagingToken/MANIFEST) As we only need manifest file to get locations for all the artifacts, passing the path to manifest file as token will remove implicit directory structure and simplify the code.

Reply:
Ok, after checking out your usage of ProxyManifest, I see what you mean.  I agree that it will resolve the concerns @jkff has about decoupling naming from storage, and allow the retrieval token to just be the manifest URI.  SGTM.